homeassistant/components/reolink/binary_sensor.py

```diff
@@ -0,0 +1,172 @@
+    PET_DETECTION_TYPE,
+    VEHICLE_DETECTION_TYPE,
+    Host,
+)
+
+from homeassistant.components.binary_sensor import (
+    BinarySensorDeviceClass,
+    BinarySensorEntity,
+    BinarySensorEntityDescription,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import ReolinkData
+from .const import DOMAIN
+from .entity import ReolinkCoordinatorEntity
+
+
+@dataclass
+class ReolinkBinarySensorEntityDescriptionMixin:
+    """Mixin values for Reolink binary sensor entities."""
+
+    value: Callable[[Host, int | None], bool]
+
+
+@dataclass
+class ReolinkBinarySensorEntityDescription(
+    BinarySensorEntityDescription, ReolinkBinarySensorEntityDescriptionMixin
+):
+    """A class that describes binary sensor entities."""
+
+    icon: str = "mdi:motion-sensor"
+    icon_off: str = "mdi:motion-sensor-off"
+    supported: Callable[[Host, int | None], bool] = lambda host, ch: True
+
+
+BINARY_SENSORS = (
+    ReolinkBinarySensorEntityDescription(
+        key="motion",
+        name="Motion",
+        device_class=BinarySensorDeviceClass.MOTION,
+        value=lambda api, ch: api.motion_detected(ch),
+    ),
+    ReolinkBinarySensorEntityDescription(
+        key=FACE_DETECTION_TYPE,
+        name="Face",
+        icon="mdi:face-recognition",
+        device_class=BinarySensorDeviceClass.MOTION,
+        value=lambda api, ch: api.ai_detected(ch, FACE_DETECTION_TYPE),
+        supported=lambda api, ch: api.ai_supported(ch, FACE_DETECTION_TYPE),
+    ),
+    ReolinkBinarySensorEntityDescription(
+        key=PERSON_DETECTION_TYPE,
+        name="Person",
+        device_class=BinarySensorDeviceClass.MOTION,
+        value=lambda api, ch: api.ai_detected(ch, PERSON_DETECTION_TYPE),
+        supported=lambda api, ch: api.ai_supported(ch, PERSON_DETECTION_TYPE),
+    ),
+    ReolinkBinarySensorEntityDescription(
+        key=VEHICLE_DETECTION_TYPE,
+        name="Vehicle",
+        icon="mdi:car",
+        icon_off="mdi:car-off",
+        device_class=BinarySensorDeviceClass.MOTION,
+        value=lambda api, ch: api.ai_detected(ch, VEHICLE_DETECTION_TYPE),
+        supported=lambda api, ch: api.ai_supported(ch, VEHICLE_DETECTION_TYPE),
+    ),
+    ReolinkBinarySensorEntityDescription(
+        key=PET_DETECTION_TYPE,
+        name="Pet",
+        icon="mdi:dog-side",
+        icon_off="mdi:dog-side-off",
+        device_class=BinarySensorDeviceClass.MOTION,
+        value=lambda api, ch: api.ai_detected(ch, PET_DETECTION_TYPE),
+        supported=lambda api, ch: api.ai_supported(ch, PET_DETECTION_TYPE),
+    ),
+    ReolinkBinarySensorEntityDescription(
+        key="visitor",
+        name="Visitor",
+        icon="mdi:bell-ring-outline",
+        icon_off="mdi:doorbell",
+        value=lambda api, ch: api.visitor_detected(ch),
```

Comment:
What's triggering this sensor? Why is this not classified as a motion sensor with that device class?

Reply:
The "visitor" sensor is the physical button on a video doorbell. So this binary sensor shows when a doorbell is beeing pressed down. (you get API webhook calls for visitor: true and visitor: false.

Reply:
Reolink calls this "visitor" in their API and in there web/windows/phone applications.
But we might want to rename this to "Doorbell press", since I find that a bit more discriptive.
@MartinHjelmare What do you think?

Reply:
> The "visitor" sensor is the physical button on a video doorbell. 

It should not be a binary sensor. Button presses are events.

Reply:
@frenck the reolink API has the "GetEvents" command that you can use to poll the status of the "visitor" that returns True/False, the ONVIF notifications for the webhook also have visitor: true and visitor: false (you get 2 notifications, 1 true and a bit later false). I think they also have logic that if the doorbell is pressed twice within a certain amount the visitor state will remain true (I actually have not tested this).

So reolink implemented it as a boolean sensor in there API.
I can make logic that compares a chached value with a new value and fire a event when the value changes from true to false, but as far as I was aware we are supposed to deliver the entities as close to the API as possible.

Reply:
Yes, button presses, that are stateless, should be custom events. Although, if the device API has a state for a feature it's ok to use that state, I'd say.

Where is the visitor state created? Is it created in the Reolink camera software or is it the PyPI library that creates the state from an event. If it's the latter I'd remove that state, as it's not a true representation of the event. If it's actually the camera that says `visitor: true` and `visitor: false` and it keeps the `true` state for longer than some milliseconds corresponding to the button push, I think it's ok to use that as a binary sensor, without device class.

Reply:
The state is created in the Reolink camera software, the API of the reolink camera (not Pypi) provides the boolean values using the GetEvents command and through there ONVIF notifications.

Reply:
The camera keeps the state for a few seconds, I have not timed it, but I think around 15 seconds or so.

Reply:
This is the reolink API response to a GetEvents request (this is raw what the camera returns over HTTP(s))
```
[{
        "cmd": "GetEvents",
        "code": 0,
        "value": {
                "ai": {
                        "dog_cat": {
                                "alarm_state": 0,
                                "support": 0
                        },
                        "face": {
                                "alarm_state": 0,
                                "support": 0
                        },
                        "people": {
                                "alarm_state": 0,
                                "support": 1
                        },
                        "vehicle": {
                                "alarm_state": 0,
                                "support": 0
                        }
                },
                "channel": 0,
                "md": {
                        "alarm_state": 0,
                        "support": 1
                },
                "visitor": {
                        "alarm_state": 1,
                        "support": 1
                }
        }
}]
```

Reply:
Ok. Then I think it's ok to keep the binary sensor for visitor without device class.

Reply:
Good, thank you!
It already does not have a device class.